src/lib.rs

```diff
@@ -57,66 +63,131 @@ pub fn dummy_endpoint() -> String {
     #[cfg(windows)]
-    pub fn incoming(self) -> Incoming {
-        Incoming { inner: NamedPipeSupport { path: self._path, handle: self._handle.remote().clone(), pipe: self.inner } }
+    pub fn incoming(self, handle: Handle) -> io::Result<Incoming> {
+        let pipe = self.inner(&handle)?;
+        Ok(
+            Incoming { inner: NamedPipeSupport { path: self.path, handle: handle.remote().clone(), pipe: pipe, security_attributes: self.security_attributes.0} }
+          )
     }
 
     /// Inner platform-dependant state of the endpoint
     #[cfg(windows)]
-    fn inner(p: &str, handle: &Handle) -> io::Result<NamedPipe> {
-        NamedPipe::new(p, handle)
+    fn inner(&self, handle: &Handle) -> io::Result<NamedPipe> {
+        extern crate mio_named_pipes;
+        use std::os::windows::io::*;
+        use miow::pipe::NamedPipeBuilder;
+
+        let raw_handle = unsafe { NamedPipeBuilder::new(&self.path)
+            .first(true)
+            .inbound(true)
+            .outbound(true)
+            .out_buffer_size(65536)
+            .in_buffer_size(65536)
+            .with_security_attributes(self.security_attributes.0)?
+            .into_raw_handle()};
+
+        let mio_pipe = unsafe { mio_named_pipes::NamedPipe::from_raw_handle(raw_handle) };
+        NamedPipe::from_pipe(mio_pipe, handle)
     }
 
     /// Inner platform-dependant state of the endpoint
     #[cfg(not(windows))]
-    fn inner(p: &str, handle: &Handle) -> io::Result<tokio_uds::UnixListener> {
-        tokio_uds::UnixListener::bind(p, handle)
+    fn inner(&self, handle: &Handle) -> io::Result<tokio_uds::UnixListener> {
+        tokio_uds::UnixListener::bind(&self.path, handle)
     }
 
+    /// Sets the security attributes of the underlying named pipes. Note that if the pointer to the
+    /// security attributes is not null, it should be valid for the lifetime of this struct and the
+    /// lifetime of the returned `Incoming` struct.
+    #[cfg(windows)]
+    pub unsafe fn set_security_attributes(&mut self, security_attributes: *mut SECURITY_ATTRIBUTES) {
+        self.security_attributes = SecurityAttrWrap(security_attributes);
+    }
+
+    /// Returns the path of the endpoint.
+    pub fn path(&self) -> &str {
+        &self.path
+    }
+
+
     /// New IPC endpoint at the given path
-    /// Endpoint ready to accept connections immediately
-    pub fn new(path: String, handle: &Handle) -> io::Result<Self> {
-        Ok(Endpoint { 
-            inner: Self::inner(&path, handle)?,
-            _path: path, 
-            _handle: handle.clone(),
-        })
+    pub fn new(path: String) -> Self {
+        Endpoint {
+            path: path,
+            #[cfg(windows)]
+            security_attributes: SecurityAttrWrap(ptr::null_mut()),
+        }
     }
 }
 
+// This is required as otherwise the pointer to the SECURITY_ATTRIBUTES does not implement Send
+#[cfg(windows)]
+struct SecurityAttrWrap ( *mut SECURITY_ATTRIBUTES );
```

Comment:
Yeah, nothing complicated, this is what I had in mind

Let's use it in `set_security_attributes` and remove `unsafe` there? 

Also, smth like this would be cool:

```
impl SecurityAttrWrap {
    pub fn unspecified() -> Self {
        SecurityAttrWrap(ptr::null_ptr());
    }

    pub unsafe fn from_raw(raw: *mut SECURITY_ATTRIBUTES) -> Self {
          SecurityAttrWrap(raw);
    }
}
```